src/fairseq2/recipes/hf/datasets.py

```diff
@@ -0,0 +1,188 @@
+
+        :param gang:
+            The gang over which to shard the dataset.
+        :param max_num_elements:
+            The maximum number of elements in each batch.
+        :param max_seq_len:
+            The maximum sequence length of each example.
+            Examples longer than this value will be cropped.
+        :param min_seq_len:
+            The minimum sequence length of each example. Examples
+            shorter than this value will be dropped.
+        :param seq_len_col:
+            The column that stores length of source sequence. Must
+            be provided if the data is to be batched by length
+        :param num_accumulate:
+            The number of batches to accumulate in each iteration.
+            Typically used with gradient accumulation during
+            training.
+        :param num_prefetch:
+            The number of batches to prefetch in background.
+        :param seed:
+            The seed to initialize the random number generators
+            used internally.
+        :param extras:
+            The extra parameters specific to the dataset
+            implementation.
+        """
+
+        # Resolve the batching strategies
+        def resolve_batching_strategy() -> int:
+            if bucket_size is None and max_num_elements is None:
+                return 0  # NO_BATCH
+            elif bucket_size is not None and max_num_elements is not None:
+                raise ValueError("Can only batch by size or by length, not both")
+            elif bucket_size is not None:
+                return 1  # BATCH_BY_SIZE
+            else:
+                assert (
+                    max_seq_len
+                ), "Max_seq_len must be specified in batch-by-length mode"
+                return 2  # BATCH_BY_LEN
+
+        batch_mode = resolve_batching_strategy()
+        if batch_mode == 0:
+            data = dataset.data.to_batches()  # list of RecordBatch
+        else:
+            data = dataset.to_list()
+        builder = read_sequence(data)
+
+        seed += 1
+
+        # Shard.
+        builder.shard(gang.rank, gang.size, allow_uneven=True)
+
+        seed += gang.rank
+
+        if batch_mode == 0:
+            # Convert RecordBatch to python dictionary
+            builder.map(lambda batch: batch.to_pydict())
+
+        else:
+            if batch_mode == 1:
+                assert bucket_size
+                builder.bucket(bucket_size=bucket_size)
+
+            if batch_mode == 2:
+
+                skip_below_min_examples = min_seq_len is not None
+                min_seq_len = min_seq_len or 1
+
+                # Bucket by audio length.
+                assert max_num_elements and max_seq_len
+                bucket_sizes = create_bucket_sizes(
+                    max_num_elements=max_num_elements,
+                    max_seq_len=max_seq_len,
+                    min_seq_len=min_seq_len,
+                    num_seqs_multiple_of=8,
```

Comment:
Is this "multiple-of" really needed for this dataset?

Reply:
ok, 8 is not a good default value.

I would still keep num_seqs_multiple_of, but now as an extra option. The point is we can also use this HuggingFace dataset wrapper to load an ASR dataset, so we should keep the same options used in the ASR use case.